### PR TITLE
IDENTITY-6525: Adding unit tests for claim metadata mgt

### PR DESCRIPTION
**Pull request by darshanasbg**
### Proposed changes in this pull request

Adding unit tests for ClaimMetadataUtils.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+    LocalClaimDTO[] localClaimDTOs;
+
+    ExternalClaim externalClaim;
+    ExternalClaim[] externalClaims;
+
+    ExternalClaimDTO externalClaimDTO;
+
+
+    @BeforeMethod
+    public void setUp() throws Exception {
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Do we need to do this for BeforeMethod?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+        LocalClaimDTO localClaimDTO1 = ClaimMetadataUtils.convertLocalClaimToLocalClaimDTO(localClaim1);
+        Assert.assertEquals(localClaimDTO1.getLocalClaimURI(), localClaim1.getClaimURI());
+
+        LocalClaimDTO localClaimDTO2 = ClaimMetadataUtils.convertLocalClaimToLocalClaimDTO(localClaim2);
+
+        Assert.assertEquals(localClaimDTO2.getLocalClaimURI(), localClaim2.getClaimURI());
+        Assert.assertEquals(localClaimDTO2.getAttributeMappings().length, localClaim2.getMappedAttributes().size());
+        Assert.assertEquals(localClaimDTO2.getClaimProperties().length, localClaim2.getClaimProperties().size());
+
+        // TODO:
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: remove todo

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    public void testConvertExternalClaimToExternalClaimDTO() throws Exception {
+
+        ExternalClaimDTO externalClaimDTO = ClaimMetadataUtils.convertExternalClaimToExternalClaimDTO(externalClaim);
+
+        Assert.assertEquals(externalClaimDTO.getExternalClaimDialectURI(), externalClaim.getClaimDialectURI());
+        Assert.assertEquals(externalClaimDTO.getExternalClaimURI(), externalClaim.getClaimURI());
+        Assert.assertEquals(externalClaimDTO.getMappedLocalClaimURI(), externalClaim.getMappedLocalClaim());
+
+        // TODO: Pass null reference
+//        ExternalClaimDTO externalClaimDTO = ClaimMetadataUtils.convertExternalClaimToExternalClaimDTO(null);
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: commented out code.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        }
+
+
+        ClaimMapping claimMapping2 = ClaimMetadataUtils.convertExternalClaimToClaimMapping(externalClaim, null, -1234);
+
+        Assert.assertEquals(claimMapping2.getClaim().getDialectURI(), externalClaim.getClaimDialectURI());
+        Assert.assertEquals(claimMapping2.getClaim().getClaimUri(), externalClaim.getClaimURI());
+
+    }
+//
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: commented out code

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/LocalClaim.java [hunk 5]
```diff
@@ -65,6 +65,9 @@ public String getMappedAttribute(String userStoreDomainName) {
     }
 
     public void setMappedAttributes(List<AttributeMapping> mappedAttributes) {
+        if (mappedAttributes == null) {
+            return;
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/LocalClaim.java]: Would it make sense to have a debug log here?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/LocalClaim.java [hunk 6]
```diff
@@ -84,6 +87,9 @@ public String getClaimProperty(String propertyName) {
     }
 
     public void setClaimProperties(Map<String, String> claimProperties) {
+        if (claimProperties == null) {
+            return;
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/LocalClaim.java]: Would it make sense to have a debug log here?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtils.java [hunk 7]
```diff
@@ -43,6 +45,10 @@
  */
 public class ClaimMetadataUtils {
 
+    private ClaimMetadataUtils() {
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtils.java]: Do we need a new line here?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+import org.testng.annotations.Test;
+import org.wso2.carbon.identity.claim.metadata.mgt.dto.*;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.AttributeMapping;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.ClaimDialect;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.ExternalClaim;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.LocalClaim;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.claim.ClaimMapping;
+
+import java.util.*;
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Let's avoid star imports :wink:

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+    }
+
+    private void setUpExternalClaims() {
+
+        String claimDialectURI = "testClaimDialectURI";
+        String externalClaimURI = "testExternalClaimURI";
+        String mappedLocalClaimURI = "testLocalClaimURI2";
+
+        externalClaim = new ExternalClaim(claimDialectURI, externalClaimURI, mappedLocalClaimURI);
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Let' remove one new line here.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+        claimPropertiesMap.put(ClaimConstants.REGULAR_EXPRESSION_PROPERTY, "^[\\S]{5,30}$");
+        claimPropertiesMap.put(ClaimConstants.DISPLAY_ORDER_PROPERTY, "1");
+        claimPropertiesMap.put(ClaimConstants.SUPPORTED_BY_DEFAULT_PROPERTY, "true");
+        claimPropertiesMap.put(ClaimConstants.REQUIRED_PROPERTY, "true");
+        claimPropertiesMap.put(ClaimConstants.READ_ONLY_PROPERTY, "true");
+        claimPropertiesMap.put(ClaimConstants.DEFAULT_ATTRIBUTE, "uid");
+
+        LocalClaim localClaim3 = new LocalClaim(localClaimURI3, attributeMappingList, claimPropertiesMap);
+
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Extra new lines

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+    @Test(dataProvider = "Authentication")
+    public void testConvertLocalClaimToClaimMapping(LocalClaim localClaim) throws Exception {
+
+        ClaimMapping claimMapping = ClaimMetadataUtils.convertLocalClaimToClaimMapping(localClaim, -1234);
+
+        Assert.assertEquals(claimMapping.getClaim().getDialectURI(), localClaim.getClaimDialectURI());
+        Assert.assertEquals(claimMapping.getClaim().getClaimUri(), localClaim.getClaimURI());
+
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Extra new lines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+            Assert.assertEquals(claimMapping.getClaim().isRequired(), Boolean.parseBoolean(claimProperties.get(
+                    ClaimConstants.REQUIRED_PROPERTY)));
+        }
+
+        if (claimProperties.containsKey(ClaimConstants.READ_ONLY_PROPERTY)) {
+            Assert.assertEquals(claimMapping.getClaim().isReadOnly(), Boolean.parseBoolean(claimProperties.get(
+                    ClaimConstants.READ_ONLY_PROPERTY)));
+        }
+
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Extra new lines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+        for (AttributeMapping attributeMapping : localClaim2.getMappedAttributes()) {
+            Assert.assertEquals(claimMapping.getMappedAttribute(attributeMapping.getUserStoreDomain()),
+                    attributeMapping.getAttributeName());
+        }
+
+
```
Comment (mefarazath) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Extra new lines

**Pull request merged**